### DISCUSSION TITLE
How to utilize multi-cores CPU for simulation

### DISCUSSION THREAD

**Discussion by Arinomo** [OpenModelica/OpenModelica#9573] [Q&A] [answered]
IIRC OpenModelica will use all available CPU cores when translating a model but used only one core to simulate it. I've notice the option "Enable parallelization of independent systems of equation" but not sure what it does or how should one used this option. Is there a way to use multi-core for one simulation? If yes, will this increase the performance of simulation(i.e. will it reduced the time needed for x s simulation)?

Background:
I try to simulate a quite complex electrical circuit and would like to find a way to improve the simulation's performance.

P.S.: If someone could give me some pointer on what to read about increasing simulation performance in general, i'd be delighted :D

**Comment by perost** [comment]
@mahge should have some info on that.

But just to clarify we do not use all available CPU cores when translating either. The compiler itself is almost entirely single-threaded since the language it's implemented in (MetaModelica) does not have any real support for parallel programming. The garbage collector in MetaModelica runs in separate threads though, so you might see more than one core being utilized due to that.

**Reply by Arinomo** [reply]
Ah, since i saw the `number of Processors` option is set to 12 (in my case) and on task-manager all the core seem to spike at compilation, i thought it use all the cores., but thanks for the clarification. :D

**Reply by adeas31** [reply]
`number of processors` is only used for parallel compilation of the generated C code.

**Reply by casella** [reply]
Regarding multi-threaded compilation, I think there are some phases in backend and codegen that carry out work in parallel, though I'm not 100% sure. For sure having garbage collection running in parallel helps a lot, and so does compiling C source files in parallel, thought the balancing between the length of the different generated C source files must still be improved.

In my experience, when compiling large models with a 20-core server, the speed up between the actual process time and the wall time can be significant (well above 2X)

Regarding parmodauto, we're starting to experiment on real-life cases right now. I'll put a note in the documentation of parmodauto that it's Linux only.

**Reply by casella** [reply]
#9581 gives more info on --parmodauto regarding the fact that it only works in Linux 😅

**Reply by perost** [reply]
> Regarding multi-threaded compilation, I think there are some phases in backend and codegen that carry out work in parallel, though I'm not 100% sure.

There are two things in the compiler that's done in parallel: parse files and generate simulation code. But beyond the trivial "launch a bunch of independent threads and wait for the result" kind of parallelization it's not really possible to do much more in MetaModelica.

**Reply by casella** [reply]
Until we have proper array-aware code generation from the new backend, that helps a lot 😅

**Comment by mahge** [comment, marked as answer]
> IIRC OpenModelica will use all available CPU cores when translating a model but used only one core to simulate it. I've notice the option "Enable parallelization of independent systems of equation" but not sure what it does or how should one used this option.

There is an experimental support for multi-core simulation. Unfortunately, it is only supported on Linux at the moment.

You can enable it by setting the flag `--parmodauto` in OMEdit (tools->options->Additional Translation Flags) or if you are using a script, by specifying `setCommandLineOptions("--parmodauto");`

>  If yes, will this increase the performance of simulation(i.e. will it reduced the time needed for x s simulation)?

This is highly dependent on the nature of you domain, structure and size of your models. The larger the model the better chance for any meaningful performance improvement.  However, even for large models, there is sometimes not enough parallelization potential depending on the structure of the model. 

You can give it a try on your models and see if you get any meaningful improvements. If there are issues with the usage, let me know and I will assist you.

**Reply by Arinomo** [reply]
I see, i did tried to set `--paramodauto=true` (on win10) and maybe that's why i didn't notice any difference. I found this option on the user guide, it would be nice if there is some indicator if an experimental function work on multiple platforms or only on some. Thanks for the reply, i'll marked it as answer to this question.

**Reply by Arinomo** [reply]
just noticed, there was an Scripting Error that the flag is unknown on the Errors tab but I've must missed it since the compiling went anyway.

**Reply by adeas31** [reply]
> You can enable it by setting the flag --parmodauto in OMEdit (tools->options->Additional Translation Flags) or if you are using a script, by specifying setCommandLineOptions("--parmodauto");

Or by simply checking `Tools->Options->Simulation->Enable parallelization of independent systems of equations (Experimental)`.

**Comment by mahge** [comment]
Also, as @perost mentioned this applies only to the simulation of the model and not the compilation/translation. 

> P.S.: If someone could give me some pointer on what to read about increasing simulation performance in general, i'd be delighted :D

You can probably get some good pointers from @casella if you provide your models or more info about them.

**Reply by Arinomo** [reply]
Providing the model would be quite difficult, but maybe i will try to create a MWE of my case and post it here.

**Reply by casella** [reply]
Please do
